cmd/nerdctl/compose_ps_linux_test.go

```diff
@@ -0,0 +1,105 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/containerd/nerdctl/pkg/testutil"
+)
+
+func TestComposePsJSON(t *testing.T) {
+	base := testutil.NewBase(t)
+	var dockerComposeYAML = fmt.Sprintf(`
+version: '3.1'
+
+services:
+  wordpress:
+    image: %s
+    ports:
+      - 8080:80
+    environment:
+      WORDPRESS_DB_HOST: db
+      WORDPRESS_DB_USER: exampleuser
+      WORDPRESS_DB_PASSWORD: examplepass
+      WORDPRESS_DB_NAME: exampledb
+    volumes:
+      - wordpress:/var/www/html
+  db:
+    image: %s
+    environment:
+      MYSQL_DATABASE: exampledb
+      MYSQL_USER: exampleuser
+      MYSQL_PASSWORD: examplepass
+      MYSQL_RANDOM_ROOT_PASSWORD: '1'
+    volumes:
+      - db:/var/lib/mysql
+
+volumes:
+  wordpress:
+  db:
+`, testutil.WordpressImage, testutil.MariaDBImage)
+
+	comp := testutil.NewComposeDir(t, dockerComposeYAML)
+	defer comp.CleanUp()
+	projectName := comp.ProjectName()
+	t.Logf("projectName=%q", projectName)
+
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "up", "-d").AssertOK()
+	defer base.ComposeCmd("-f", comp.YAMLFullPath(), "down", "-v").Run()
+
+	assertHandler := func(svc string, count int, fields ...string) func(stdout string) error {
+		return func(stdout string) error {
+			// 1. check json output can be unmarshalled back to printables.
+			var printables []composeContainerPrintable
+			if err := json.Unmarshal([]byte(stdout), &printables); err != nil {
+				return fmt.Errorf("[service: %s]failed to unmarshal json output from `compose ps`: %s", svc, stdout)
+			}
+			// 2. check #printables matches expected count.
+			if len(printables) != count {
+				return fmt.Errorf("[service: %s]unmarshal generates %d printables, expected %d: %s", svc, len(printables), count, stdout)
+			}
+			// 3. check marshalled json string has all expected substrings.
+			for _, field := range fields {
+				if !strings.Contains(stdout, field) {
+					return fmt.Errorf("[service: %s]marshalled json output doesn't have expected string (%s): %s", svc, field, stdout)
+				}
+			}
+			return nil
+		}
+	}
+
+	// check other formats are not supported
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "ps", "--format", "yaml").AssertFail()
```

Comment:
Not sure why this line will fail docker-compatible CI test 🤔 I checked docker code and tested locally, docker does return error (not 0) here.

```shell
=== RUN   TestComposePsJSON
    compose_ps_linux_test.go:63: projectName="nerdctl-compose-test2770844775"
    compose_ps_linux_test.go:90: assertion failed: res.ExitCode is 0
--- FAIL: TestComposePsJSON (13.36s)
```

```
$ docker compose ps --format yaml
format value "yaml" could not be parsed: parsing failed
$ echo $?
1
$ sudo nerdctl compose ps --format yaml
FATA[0000] unsupported format yaml, supported formats are: [json]
$ sudo nerdctl compose echo $?
1
```
docker-compose code: [1](https://github.com/docker/compose/blob/v2/cmd/compose/ps.go#L143), [2](https://github.com/docker/compose/blob/v2/cmd/formatter/formatter.go#L70)